### DISCUSSION TITLE
Choppy Audio for A SD to PWM output

### DISCUSSION THREAD

**Discussion by ChrisComos** [pschatzmann/arduino-audio-tools#1540] [Q&A] [answered]
Currently getting choppy audio when attempting to play a WAV file from an SD card on a Seeeduino XIAO Sense nRF52840
My wiring seems correct since I can play the memory to PWM example just fine, running the PWM signal to a PAM8302A Class D amp and getting decent sound.

I have attempted to change buffer sizes for the player, pwm stream, and config but nothing seems to work for me which makes me think my SD card is too slow.

I figure this is an implementation error on my end since this is my first time using this board/chip and library.
Any advice or insight would be greatly appreciated.

Here is my current code
```C
#include "AudioTools.h"
#include "AudioLibs/AudioSourceSDFAT.h"
#include "AudioCodecs/CodecWAV.h"


const char *startFilePath="/";
const char* ext="wav";
AudioSourceSDFAT source(startFilePath, ext,2);
AudioInfo info(8000, 1, 16);
WAVDecoder wav;
PWMAudioOutput pwm;          // PWM output 
AudioPlayer player(source, pwm, wav);

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  Serial.begin(115200);
  while(!Serial){}
  Serial.print("Starting");
  AudioLogger::instance().begin(Serial, AudioLogger::Info);
  player.setMetadataCallback(printMetaData);
  

  auto config = pwm.config();
  config.setAudioInfo(info);
  pwm.begin(config);
  
  source.begin();
  
  // setup player
  wav.begin();

  source.setFileFilter("*test*");
  player.begin();
}

void loop() {
  player.copy();
}
```

**Comment by pschatzmann** [comment, marked as answer]
You can define the clock speed in the constructor of the AudioSourceSDFAT and don't forget to set the log level to Warning

**Reply by ChrisComos** [reply]
I changed the log level to Warning instead of Info and I also set the constructor to 
```C
AudioSourceSDFAT source(startFilePath, ext,2,50);
```
There was no, noticeable change in the output. Could there be a buffer that I missed?

**Comment by pschatzmann** [comment]
Have a look at PWMConfig.
What is the sample rate in your wav files ?

**Reply by ChrisComos** [reply]
16000 hz, mono, 8bit PCM using audio according to Audacity
Fixed this, since I must've gotten them mixed 
```C
AudioInfo info(16000, 1, 8);
``` 
Still no change in audio.

**Reply by pschatzmann** [reply]
Did you try to increase the buffer in PWMConfig ?
If this does not help I would suggest to try to change the wav file to use a recording with 8000 instead of 16000

**Reply by ChrisComos** [reply]
Yeah that worked, It was the file that needed changed to 8000 and it now works. Actually I was able to get it to 11025. I will see if there is anything else I can do to increase that to get the best quality I can.
Thanks for your help!
